server/db-patches/patch-22-23.sql

```diff
@@ -0,0 +1,22 @@
+-- pg-patcher runs migrations in transactions. However, ALTER TYPE ... ADD VALUE
+-- cannot be executed inside a transaction block. We will swap out the enum type
+-- with a new one instead.
+ALTER TYPE shot_block_type RENAME TO old_shot_block_type;
+CREATE TYPE shot_block_type AS ENUM ('none', 'dmca', 'abuse', 'usererror', 'watchdog');
+ALTER TABLE data ALTER COLUMN block_type DROP DEFAULT;
+ALTER TABLE data ALTER COLUMN block_type TYPE shot_block_type USING block_type::text::shot_block_type;
```

Comment:
This isn't great. Basically it's recreating the entire column by converting the existing value to text and then that text to the new enum type. Because the column is rewritten, the whole row is rewritten, and therefore the whole table. The database has a row size of around 500 bytes, and with 4.2e7 rows, it will take a while. PostgreSQL will take an exclusive lock on the whole table during the migration, so while those 21 GB are being written to disk, that table won't be usable. There are alternatives we could discuss.

Also I might be wrong as I'm certainly not a postgres expert, but testing on my laptop does seem to indicate the full table rewrite is happening.

Reply:
Actually, I don't think we need to discuss this, I found out since PostgreSQL 9.1 there's a strictly better alternative. Since 9.1, you can simply `ALTER TYPE enum ADD VALUE 'foo';`.

Here's a quick benchmark of what I propose compared to what this PR does: https://gist.github.com/autrilla/3837ed08581a30132594e24a7f640a4c

As you can see, `ADD VALUE` is very fast, and this PR indeed rewrites the whole table.

Reply:
@autrilla The original patch did use `ALTER TYPE ... ADD VALUE`; see the comment at https://github.com/mozilla-services/screenshots/pull/4417/files#diff-23625b9e1cf673be5214f52029f9531bR1 on why that was changed.  There was also [some discussion in IRC](https://mozilla.logbot.info/screenshots/20180515#c14759281) on potential approaches.

Reply:
My bad, I missed that. Let's discuss it on Monday then!

Reply:
We might want to move away from ENUMs if this list is going to change or keep growing.

Reply:
We could use a natural foreign key to a table of block types or a check constraint instead.

Reply:
Yeah, I agree, though we still need to migrate away from the current implementation without downtime

Reply:
@chenba it's dirty, but what does pg-patcher do if we were to start our migration SQL with

```sql
COMMIT;
ALTER TYPE ... ADD VALUE;
BEGIN;
```

If it's supported, it's probably the simplest way. Other alternatives I can think of:

1. Do what @oremj said, use a table of block types, add a new column that uses that table, and backfill. The application will have to read and write to both columns for the time being.
2. Add the value to the database manually.
3. Patch pg-patcher so we can have migrations that run outside a transaction.

My opinion is that backfilling is going to be very cumbersome.

Reply:
@autrilla Postgresql would throw the error: "ALTER TYPE ... ADD cannot be executed from a function or multi-command string"